### PR TITLE
Issue-147: Add encryption to sensor comms

### PR DESCRIPTION
**Pull request by drumwhizz**

### CODE REVIEW

**Review by RobHumphris — COMMENTED** [review]
10 is a magic number, it would be better to define the RecorderEventHeader size with the other constants, and then we would be best served to define all the other Header sizes and test them in each NewXXX function

File: ublox-bluetooth-commands.go [hunk 1]
```diff
@@ -164,6 +164,11 @@ func (ub *UbloxBluetooth) ConnectToDevice(address string, onConnect DeviceEvent,
 	var d []byte
 	var cr *ConnectionReply
 
+	err = ub.ConfigureSecurity(oobTemporaryKey, securityEnabledOutOfBand)
+	if err != nil {
+		return err
+	}
+
```
Comment (drumwhizz) [inline ublox-bluetooth-commands.go]: Martin has asked me to make this configurable so we are not finished yet on this review

Comment (drumwhizz) [reply ublox-bluetooth-commands.go]: Sensibly he said the same as you, Encryption is a pain and things will go wrong before they go right..

File: event-types.go [hunk 2]
```diff
@@ -109,34 +209,34 @@ func newErrorEvent(b []byte) *VehEvent {
...
 	eb.SensorEvent = &VehSensorEvent{
-		float32(binary.LittleEndian.Uint16(b[10:12]) / 4),
-		float32(binary.LittleEndian.Uint16(b[12:14]) / 1000),
+		float32(binary.LittleEndian.Uint16(b[esTemperatureOffset.Start():esTemperatureOffset.End()]) / 4),
+		float32(binary.LittleEndian.Uint16(b[esBatteryOffset.Start():esBatteryOffset.End()]) / 1000),
 		make([]byte, l-14),
 	}
-	copy(eb.SensorEvent.Other, b[13:l-1])
+	copy(eb.SensorEvent.Other, b[esOtherOffset.Start()-1:l-1])
 	return &eb
```
Comment (drumwhizz) [inline event-types.go:240]: I am not sure WTF this line of code is doing, it does not look correct to me, but I have reproduce the behaviour in the new code. Can you please double check this?

Comment (RobHumphris) [reply event-types.go:240]: Its copying the contents of the slice into the Other field
What happens is that some unknown data (`"other": "X2orPw==",`) ends up getting written to the database

**Review by RobHumphris — APPROVED** [review]

File: ublox-bluetooth.go [hunk 3]
```diff
@@ -127,6 +128,17 @@ func (btd *BluetoothDevices) CloseAll() error {
 	})
 }
 
+// EncrptComms Switch bluetooth comms encryption on/off
+func (btd *BluetoothDevices) EncrptComms(enabled bool) error {
+	return btd.ForEachDevice(func(ub *UbloxBluetooth) error {
```
Comment (drumwhizz) [inline ublox-bluetooth.go:138]: Spelling mistake

File: ublox-bluetooth-commands.go [hunk 4]
```diff
@@ -47,7 +48,22 @@ func (ub *UbloxBluetooth) getSerialNumber() (string, error) {
...
-		return string(lines[0]), nil
+		return strings.Trim(string(lines[0]), "\""), nil
+	}
+
+	return "Unknown", nil
+}
+
+// getLocalName retrieves the local name of the dongle (which will be set to the 8power 16 digit serial number)
+func (ub *UbloxBluetooth) getLocalName() (string, error) {
+	sn, err := ub.writeAndWait(GetLocalName(), true)
```
Comment (drumwhizz) [inline ublox-bluetooth-commands.go:59]: This command is timing out. I am pretty sure the response string is correct but it does not detect the response returned correctly

File: ublox-edm-commands.go [hunk 5]
```diff
@@ -76,29 +76,33 @@ func (ub *UbloxBluetooth) ParseEDMMessage(msg []byte) error {
...
-	case StartEvent:
-		ub.StartEventReceived = true
-	case ATConfirmation:
-		switch data[0] {
-		case '+':
+	m := bytes.TrimPrefix(msg[2:len(msg)-1], []byte(newline))
+	m = bytes.TrimSuffix(m, []byte(newline))
+	dataLines := bytes.Split(m, []byte(newline))
+	// data := removeNewlines(msg[2 : len(msg)-1])
+	for _, data := range dataLines {
```
Comment (RobHumphris) [inline ublox-edm-commands.go:83]: nice and clean

**Review by RobHumphris — APPROVED** [review]

**Pull request merged**